From 247256873513e9487389a72a1ea70dd792dc4644 Mon Sep 17 00:00:00 2001
From: Nicolas Pitre <npitre@baylibre.com>
Date: Fri, 19 May 2023 12:00:57 -0400
Subject: [PATCH] kernel: allow for arch specific processing within busy
 loops

Give architectures that need it the ability to perform special checks
while e.g. waiting for a spinlock to become available.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>
Upstream-Status: Backport [main: aec10d465dbf91296c9999be199b3e2446315fb6]
---
 include/zephyr/spinlock.h           |  1 +
 include/zephyr/sys/arch_interface.h | 10 ++++++++++
 kernel/idle.c                       |  8 ++++++++
 3 files changed, 19 insertions(+)

diff --git a/include/zephyr/spinlock.h b/include/zephyr/spinlock.h
index 2ba01a70ba3c..2687f0227a67 100644
--- a/include/zephyr/spinlock.h
+++ b/include/zephyr/spinlock.h
@@ -153,6 +153,7 @@ static ALWAYS_INLINE k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
 
 #ifdef CONFIG_SMP
 	while (!atomic_cas(&l->locked, 0, 1)) {
+		arch_spin_relax();
 	}
 #endif
 
diff --git a/include/zephyr/sys/arch_interface.h b/include/zephyr/sys/arch_interface.h
index 3b9c0ef4a42c..2053a7b2287a 100644
--- a/include/zephyr/sys/arch_interface.h
+++ b/include/zephyr/sys/arch_interface.h
@@ -1404,6 +1404,16 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
 
 #endif /* CONFIG_PCIE_MSI_MULTI_VECTOR */
 
+/**
+ * @brief Perform architecture specific processing within spin loops
+ *
+ * This is invoked from busy loops with IRQs disabled such as the contended
+ * spinlock loop. The default implementation is a weak function that calls
+ * arch_nop(). Architectures may implement this function to perform extra
+ * checks or power management tricks if needed.
+ */
+void arch_spin_relax(void);
+
 #ifdef __cplusplus
 }
 #endif /* __cplusplus */
diff --git a/kernel/idle.c b/kernel/idle.c
index 75cf75242d7e..2390740c60f8 100644
--- a/kernel/idle.c
+++ b/kernel/idle.c
@@ -104,3 +104,11 @@ void idle(void *unused1, void *unused2, void *unused3)
 #endif
 	}
 }
+
+void __weak arch_spin_relax(void)
+{
+	__ASSERT(!arch_irq_unlocked(arch_irq_lock()),
+		 "this is meant to be called with IRQs disabled");
+
+	arch_nop();
+}
-- 
2.34.1

